javaparser-core/src/main/java/com/github/javaparser/printer/PrettyPrintVisitor.java

```diff
@@ -112,20 +150,29 @@
 import com.github.javaparser.ast.type.WildcardType;
 import com.github.javaparser.ast.visitor.Visitable;
 import com.github.javaparser.ast.visitor.VoidVisitor;
+import com.github.javaparser.printer.configuration.PrettyPrinterConfiguration;
 
 /**
  * Outputs the AST as formatted Java source code.
- *
+ * This class is no longer acceptable to use because it is not sufficiently configurable and it is too tied to a specific implementation
```

Comment:
In my opinion all the pretty printing and lexical preservation stuff should be thrown out, and replaced by:
- a library of microformatters (formatting rules) that will format one specific case (like "exactly one space between parameter list and opening brace `a() {`" that work on specific nodes and modify the underlying tokenstream.) which can be selected and configured by the user.
- the AST nodes modifying the underlying tokenstream when nodes and properties are changed/added/removed, after which they will call the user's microformatters.

That would remove an incredible amount of unnecessary complexity currently in JP, and will solve more or less all current lexical preservation and pretty printing issues.

This idea is inspired by Intellij Idea's architecture. Well, copied really ;-)

Reply:
This might sound harsh, I know, but think of it like this: it would be really interesting to figure out how to implement this correctly and see it come to life! It is much more rewarding than continuing on the current path. The Pretty Printer visitor approach will never be fun to work on, and will stay very brittle. Just look at all the weirdness that had to be introduced to get stream method calls (map, filter) to print vertically aligned. The lexical preserving printer, an impressive amount of work, has always had trouble because it is a separate component, and keeps its own data about nodes, about token, about node types, which is nowadays all available in the AST and metamodel itself. It always has to sit in its own little corner with its own information, receiving *almost* enough information through the AST listeners to do its work. Putting lexical preservation straight in the middle of the `Node`s will be a big relief!

Reply:
(For completeness: other major improvements to JP would be to move to Antlr, and to rewrite the symbol solver to use a single pass over the whole project instead of reasoning from the inside out)

Reply:
> That would remove an incredible amount of unnecessary complexity currently in JP,

In my opinion microformatters adds another level of complexity which resides in the management of a multitude of formatting rules because the concept of formatting and especially of preserving the current format is a very complex subject.

Reply:
Alright, I guess you're not really interested in my review. I'll go spend my time elsewhere :-)

Reply:
No No @matozoid i misspoke. In fact I realize that I should not have to get into this development without sharing with the team a vision of what needed to be done. I arrived last on this project and I have no intention of causing trouble in this team. I was just giving personal advice, you don't have to agree with me. If the vision on this feature is to set up a micro-formatting logic I have nothing to say except that if I had known it earlier I would not have wasted time doing the work I did. Personally I don't feel able to get into microformatting so I'll leave the topic to you and I'll just fix a few bugs.